Moving final scriptSig construction from CombineSignatures to ProduceSignature (PSBT signer logic)

achow101 opened the pull request:
Currently CombineSignatures is used to create the final scriptSig or an input. However ProduceSignature is capable of doing this itself. Using both CombineSignatures and ProduceSignature results in code duplication which is unnecessary.

To move the scriptSig construction to ProduceSignatures, the SignatureData class contains two maps to hold pubkeys mapped to signatures, and script ids mapped to scripts. DataFromTransaction is extended to be able to extract signatures, their public keys, and scripts from existing ScriptSigs.

The SignaureData are then passed down to SignStep which can use the aforementioned maps to get the signatures, pubkeys, and scripts that it needs, falling back to the actual SigningProvider and SignatureCreator if the data are not available in the SignatureData.

Additionally, Sign1 and SignN have been removed and their functionality inlined into SignStep since Sign1 is really just a wrapper around CreateSig.

Since ProduceSignature can produce the final scriptSig or scriptWitness by using SignatureData which has extracted data from the transaction, CombineSignatures is unnecessary as ProduceSignature is able to replicate all of CombineSignatures' functionality.

This also furthers BIP 174 support and begins moving towards a BIP 174 style backend.

The tests have also been updated to use the new combining methodology.

achow101 committed:
Replace CombineSignatures with ProduceSignature

Instead of using CombineSignatures to create the final scriptSig or
scriptWitness of an input, use ProduceSignature itself.

To allow for ProduceSignature to place signatures, pubkeys, and scripts
that it does not know about, we pass down the SignatureData to SignStep
which pulls out the information that it needs from the SignatureData.

achow101 committed:
Remove CombineSignatures and replace tests

Removes CombineSignatures and replaces its use in tests with
ProduceSignature to test the same behavior for ProduceSignature.

sipa commented:
Here is an extra simplification we discussed IRL: https://github.com/sipa/bitcoin/commits/achow_sigdata-partial-sigs

achow101 committed:
Inline Sign1 and SignN

Sign1 and SignN are kind of redundant so remove them and inline their
behavior into SignStep

achow101 commented on src/script/sign.h:
```diff
@@ -56,9 +56,12 @@ extern const BaseSignatureCreator& DUMMY_SIGNATURE_CREATOR;
 struct SignatureData {
     CScript scriptSig;
     CScriptWitness scriptWitness;
+    std::map<CPubKey, std::vector<unsigned char>> signatures;
```
`GetPubKey` won't work if these are `CKeyID`s

sipa replied:
Ah, good point. You could have an `std::map<CKeyID, std::pair<CPubKey, std::vector<unsigned char>>>` though, which stores both full public key and signature.

achow101 replied:
Done

achow101 commented:
I have included squashed in @sipa's suggested commits.

sipa commented on src/script/sign.cpp:
```diff
@@ -152,7 +180,7 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
         // and then the serialized subscript:
         subscript = CScript(result[0].begin(), result[0].end());
-        solved = solved && SignStep(provider, creator, subscript, result, whichType, SigVersion::BASE) && whichType != TX_SCRIPTHASH;
+        solved = solved && SignStep(provider, creator, subscript, result, whichType, SigVersion::BASE, sigdata) && whichType != TX_SCRIPTHASH;
```
If you'd introduce a `sigdata.scripts.emplace(CScriptID(subscript), subscript);` here (and similarly below for the `TX_WITNESS_V0_SCRIPTHASH` case), you get (I think) all functionality you need for operating on `SignatureData` objects without the need for the legacy partial signatures.

achow101 replied:
Done

sipa commented:
You're missing the `DUMMY_SIGNING_PROVIDER` instance.

sipa commented on src/script/sign.h:
```diff
@@ -81,4 +84,14 @@ void UpdateInput(CTxIn& input, const SignatureData& data);
  * Solvability is unrelated to whether we consider this output to be ours. */
 bool IsSolvable(const SigningProvider& provider, const CScript& script);
 
+class SignatureExtractorChecker : public MutableTransactionSignatureChecker
```
Oh of course; I just mean that you pass in a `BaseSignatureChecker*` which is stored inside. You can then call its CheckSig member function (which is public).

achow101 replied:
Ah, right. Done.

achow101 commented on src/script/sign.h:
```diff
@@ -56,9 +56,12 @@ extern const BaseSignatureCreator& DUMMY_SIGNATURE_CREATOR;
     CScript scriptSig;
     CScriptWitness scriptWitness;
+    std::map<CPubKey, std::vector<unsigned char>> signatures;
+    std::map<uint160, CScript> scripts;
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -184,12 +184,101 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
+
+    // Get scripts
+    txnouttype script_type;
+    std::vector<std::vector<unsigned char> > solutions;
```
Fixed

achow101 commented on src/script/sign.h:
```diff
@@ -56,9 +56,12 @@ extern const BaseSignatureCreator& DUMMY_SIGNATURE_CREATOR;
 struct SignatureData {
     CScript scriptSig;
```
Done

achow101 commented on src/bitcoin-tx.cpp:
```diff
@@ -648,10 +648,8 @@ static void MutateTxSign(CMutableTransaction& tx, const std::string& flagStr)
         // Only sign SIGHASH_SINGLE if there's a corresponding output:
         if (!fHashSingle || (i < mergedTx.vout.size()))
-            ProduceSignature(keystore, MutableTransactionSignatureCreator(&mergedTx, i, amount, nHashType), prevPubKey, sigdata);
+            ProduceSignature(SignatureDataSigningProvider(&sigdata, &keystore), SignatureDataSignatureCreator(&sigdata, &mergedTx, i, amount, nHashType), prevPubKey, sigdata);
```
Done

achow101 commented:
> You're missing the `DUMMY_SIGNING_PROVIDER` instance.

Fixed

sipa reviewed (commented):
utACK 64f4a710bbddace7ee51ff6cfbc19f039e7739fa, only nits remaining.

achow101 commented on src/script/sign.cpp:
```diff
@@ -33,27 +33,33 @@ bool MutableTransactionSignatureCreator::CreateSig(const SigningProvider& provid
+    // Look for scripts in SignatureData
+    auto mi = sigdata->scripts.find(scriptid);
+    if (mi != sigdata->scripts.end()) {
+        script = (*mi).second;
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -33,27 +33,33 @@ bool MutableTransactionSignatureCreator::CreateSig(const SigningProvider& provid
     }
-    return nSigned==nRequired;
+    // Look for pubkey in all partial sigs
+    for (auto& it : sigdata->signatures) {
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -33,27 +33,33 @@ bool MutableTransactionSignatureCreator::CreateSig(const SigningProvider& provid
 }
 
-static bool SignN(const SigningProvider& provider, const std::vector<valtype>& multisigdata, const BaseSignatureCreator& creator, const CScript& scriptCode, std::vector<valtype>& ret, SigVersion sigversion)
+static bool GetPubKey(const SigningProvider* provider, const SignatureData* sigdata, const CKeyID &address, CPubKey& vchPubKeyOut)
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -81,37 +88,56 @@ static bool SignStep(const SigningProvider& provider, const BaseSignatureCreator
-        }
+        if (!creator.CreateSig(provider, sig, keyID, scriptPubKey, sigversion)) return false;
+        ret.push_back(std::move(sig));
+        CPubKey vch;
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -298,89 +247,112 @@ struct Stacks
+        // Get the redeemScript
+        CScript redeem_script(stack.script.back().begin(), stack.script.back().end());
+        data.scripts.emplace(CScriptID(redeem_script), redeem_script);
+        next_script = redeem_script;
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -298,89 +247,112 @@ struct Stacks
+        // Get the witnessScript
+        CScript witness_script(stack.witness.back().begin(), stack.witness.back().end());
+        data.scripts.emplace(CScriptID(witness_script), witness_script);
+        next_script = witness_script;
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -81,37 +87,56 @@ static bool SignStep(const SigningProvider& provider, const BaseSignatureCreator
+            if (it != sigdata.signatures.end()) {
+                sig = it->second.second;
+            } else {
+                if (!creator.CreateSig(provider, sig, pubkey.GetID(), scriptPubKey, sigversion)) sig.clear();
```
done

achow101 commented on src/script/sign.cpp:
```diff
@@ -298,89 +246,112 @@ struct Stacks
+        // Get witnessScript type
+        Solver(next_script, script_type, solutions);
+        stack.witness.pop_back();
+        stack.script = stack.witness;
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -298,89 +246,112 @@ struct Stacks
-            result.witness.push_back(valtype(pubKey2.begin(), pubKey2.end()));
-            return result;
+            for (unsigned int i = 0; i < num_pubkeys; i++)
+            {
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -298,89 +246,112 @@ struct Stacks
+            {
+                const valtype& pubkey = solutions[i+1];
+                if (data.signatures.count(CPubKey(pubkey).GetID()))
+                    continue; // Already got a sig for this pubkey
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -298,89 +246,112 @@ struct Stacks
+                    continue; // Already got a sig for this pubkey
+
+                if (checker.CheckSig(sig, pubkey, next_script, sigversion))
+                {
```
Done

achow101 commented on src/script/sign.h:
```diff
@@ -53,12 +55,18 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
 
     SignatureData() {}
     explicit SignatureData(const CScript& script) : scriptSig(script) {}
+    void UpdateWithSignatureData(SignatureData sigdata);
```
Done

achow101 commented on src/test/transaction_tests.cpp:
```diff
@@ -629,7 +629,10 @@ BOOST_AUTO_TEST_CASE(test_witness)
     CheckWithFlag(output2, input2, 0, false);
     BOOST_CHECK(*output1 == *output2);
-    UpdateInput(input1.vin[0], CombineSignatures(output1->vout[0].scriptPubKey, MutableTransactionSignatureChecker(&input1, 0, output1->vout[0].nValue), DataFromTransaction(input1, 0, output1->vout[0]), DataFromTransaction(input2, 0, output1->vout[0])));
+    sigdata = DataFromTransaction(input1, 0, output1->vout[0]);
```
Done

sipa commented:
utACK 862c0d1749e28c46fd7c5fb951ac41f6d5dfad06

Perhaps something to add to the rationale: by having all processing be done by `ProduceSignature` (which now can take existing partial signatures as input), the model much more closely matches BIP 174 procedures. The changes here should make it possible for the BIP174 implementation to largely reuse the existing signing code.

achow101 commented:
I added a helper function for `createSig` which handles the fetching and placing of signatures from the SignatureData.

achow101 commented:
Removed the `newsigs` map from the first commit. Also moved some changes around to commits where they actually made sense to be in.

sipa commented:
Perhaps this PR can be renamed to "PSBT signer logic"; it's perhaps not clear right now that in addition to being a code simplification, this also effectively reduces the PSBT PR to serialization + RPC implementations.

achow101 commented on src/script/sign.cpp:
```diff
@@ -181,15 +217,104 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
+        assert(solutions.size() > 1);
+        unsigned int num_pubkeys = solutions.size()-2;
+        for (const valtype& sig : stack.script) {
+            for (unsigned int i = 0; i < num_pubkeys; i++) {
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -181,15 +217,104 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
+        for (const valtype& sig : stack.script) {
+            for (unsigned int i = 0; i < num_pubkeys; i++) {
+                const valtype& pubkey = solutions[i+1];
+                if (data.signatures.count(CPubKey(pubkey).GetID())) continue; // Already got a sig for this pubkey
```
Done

achow101 commented:
Updated the title to include "PSBT signer logic".

MeshCollider commented:
Concept ACK

sipa reviewed (commented):
utACK 1d5c2bbaba8e3b66195acefe37053d9048c81ecf

Some tiny nits only, feel free to ignore.

achow101 commented on src/script/sign.cpp:
```diff
@@ -169,15 +171,106 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
+                const valtype& pubkey = solutions[i+1];
+                // We either have a signature for this pubkey, or we have found a signature and it is valid
+                if (data.signatures.count(CPubKey(pubkey).GetID()) || checker.CheckSig(sig, pubkey, next_script, sigversion)) {
+                    last_success_key = i;
```
Done

achow101 commented on src/script/sign.cpp:
```diff
@@ -187,6 +280,17 @@ void UpdateInput(CTxIn& input, const SignatureData& data)
+        *this = std::move(sigdata);
+        return;
+    }
+    scripts.insert(sigdata.scripts.begin(), sigdata.scripts.end());
```
Done

sipa commented:
utACK 3d2cbe67e413e8de0b7144280a19bb348d1a76ca. Only change is addressing my last nits.

achow101 commented:
I rebased this and have also updated the SignatureData to match the planned changes for BIP 174.

sipa commented:
utACK 33b85bb01f71d6da7cde39eb5cf42dd581de14b6.

instagibbs commented on src/script/sign.h:
```diff
@@ -53,12 +53,19 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
+    CScript redeem_script; // The redeemScript (if any) for the input
+    CScript witness_script; // The witnessScript (if any) for the input
+    CScriptWitness scriptWitness; // The scriptWitness of an input. Contains complete signatures or the traditional partial signatures format
+    std::map<CKeyID, SigPair> signatures; // BIP 174 style partial signatures for the input. May contain complete signatures
```
may want to rephrase or define "complete signatures". I assume you mean it in the PSBT sense.

achow101 replied:
Done

instagibbs commented on src/script/sign.h:
```diff
@@ -53,12 +53,19 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
 struct SignatureData {
-    CScript scriptSig;
-    CScriptWitness scriptWitness;
+    bool complete = false; // Stores whether the scriptSig and scriptWitness are complete
```
Since we're heavily augmenting this struct could we get a description of the struct's purpose in a comment above it?

achow101 replied:
Done

instagibbs commented on src/script/sign.h:
```diff
@@ -72,7 +79,7 @@ bool SignSignature(const SigningProvider &provider, const CTransaction& txFrom,
 SignatureData CombineSignatures(const CScript& scriptPubKey, const BaseSignatureChecker& checker, const SignatureData& scriptSig1, const SignatureData& scriptSig2);
 
 /** Extract signature data from a transaction, and insert it. */
```
"from a transaction input"?

achow101 replied:
Done

instagibbs commented on src/script/sign.h:
```diff
@@ -53,12 +53,19 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
+    bool complete = false; // Stores whether the scriptSig and scriptWitness are complete
+    CScript scriptSig; // The scriptSig of an input. Contains complete signatures or the traditional partial signatures format
+    CScript redeem_script; // The redeemScript (if any) for the input
+    CScript witness_script; // The witnessScript (if any) for the input
```
witness script vs script witness.... let's explain this one as well.

achow101 replied:
Done

instagibbs commented on src/script/sign.cpp:
```diff
@@ -169,15 +171,106 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
+};
+}
+
+// Extracts signatures and scripts from scriptSigs. Please do not extend this, use PSBT instead
```
May want to note it won't extract scripts(redeem script and witness script) if the signatures are already complete

instagibbs replied:
Also, I presume this function is only dealing with multisigs, since it doesn't need to know any special scripts to otherwise sign or extract signatures for pubkey-based outputs?

achow101 replied:
Done

instagibbs commented on src/script/sign.cpp:
```diff
@@ -169,15 +171,106 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
+
+    // Get signatures
+    MutableTransactionSignatureChecker tx_checker(&tx, nIn, txout.nValue);
+    SignatureExtractorChecker checker(&data, &tx_checker);
```
suggestion: rename this `extractor_checker` to make later code more clear. Final code block looks like it doesn't change any state if reader thinks it's a different checker.

achow101 replied:
Done

instagibbs commented:
Comments on just making it clearer to readers of the code what's going on.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/13425/commits/95f5a38dd0da039e34aeb805c80e38d3374b9077

achow101 commented:
Addressed @promag's comments from https://github.com/bitcoin/bitcoin/pull/13557#pullrequestreview-132751570

sipa commented:
reutACK a0982864893db999484850282822472cce16913c

Empact commented on src/script/sign.h:
```diff
@@ -53,12 +53,22 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
+    CScript redeem_script; // The redeemScript (if any) for the input
+    CScript witness_script; // The witnessScript (if any) for the input. witnessScripts are used in P2WSH outputs.
+    CScriptWitness scriptWitness; // The scriptWitness of an input. Contains complete signatures or the traditional partial signatures format. scriptWitness is part of a transaction input per BIP 144.
+    std::map<CKeyID, SigPair> signatures; // BIP 174 style partial signatures for the input. May contain all signatures necessary for producing a final scriptSig or scriptWitness.
```
nit: these comments could be doxygenized, e.g. https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html#memberdoc

achow101 replied:
Done

Empact commented on src/script/sign.h:
```diff
@@ -21,11 +21,13 @@ class SigningProvider
-    virtual bool GetKey(const CKeyID &address, CKey& key) const =0;
+    virtual bool GetCScript(const CScriptID &scriptid, CScript& script) const { return false; }
+    virtual bool GetPubKey(const CKeyID &address, CPubKey& pubkey) const { return false; }
+    virtual bool GetKey(const CKeyID &address, CKey& key) const { return false; }
```
If you leave this pure virtual, it will continue to require that all inheritors implement the associated functions. With this change there's no such forward requirement.

Empact replied:
The theory is that if this stays an abstract base class, any future descendant are more likely to be correct & intentional if the compiler ensures that each of these methods is defined. It’s a more defensive and theoretical concern than a practical one, but I also think it’s cleaner if a dummy class does not influence the base class for all of the other classes.

Empact commented on src/script/sign.cpp:
```diff
@@ -81,30 +59,40 @@ static bool SignStep(const SigningProvider& provider, const BaseSignatureCreator
+        ret.push_back(std::move(sig));
+        return true;
+    case TX_PUBKEYHASH: {
         keyID = CKeyID(uint160(vSolutions[0]));
```
nit: This is the only use of `keyID` now, so you could declare it in this block.

achow101 replied:
Done

Empact commented on src/script/sign.h:
```diff
@@ -53,12 +53,22 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
+    CScript redeem_script; ///< The redeemScript (if any) for the input
+    CScript witness_script; ///< The witnessScript (if any) for the input. witnessScripts are used in P2WSH outputs.
+    CScriptWitness scriptWitness; ///< The scriptWitness of an input. Contains complete signatures or the traditional partial signatures format. scriptWitness is part of a transaction input per BIP 144.
+    std::map<CKeyID, SigPair> signatures; ///< BIP 174 style partial signatures for the input. May contain all signatures necessary for producing a final scriptSig or scriptWitness.
```
This is the same as CryptedKeyMap in keystore.h

sipa replied:
No it's not. This stores signatures, not keys.

Empact replied:
To be clear, the type definitions are the same:

`typedef std::map<CKeyID, std::pair<CPubKey, std::vector<unsigned char> > > CryptedKeyMap;`

sipa replied:
Ah, I see.

I generally don't think types should be identical because their construction is the same. The contents has a very different meaning, so having separate names makes sense.

Empact commented on src/script/sign.h:
```diff
@@ -53,12 +53,22 @@ class MutableTransactionSignatureCreator : public BaseSignatureCreator {
 
     SignatureData() {}
     explicit SignatureData(const CScript& script) : scriptSig(script) {}
+    void MergeSignatureData(SignatureData sigdata);
```
It's a bit strange that this commit adds MergeSignatureData but doesn't use it.

achow101 replied:
Moved to the commit where it is used.

sipa replied:
The implementation is now in the next commit, but the definition is still in "Make SignatureData able to store signatures and scripts".

achow101 replied:
Moved the definition.

sipa commented on src/script/sign.h:
```diff
@@ -81,4 +91,15 @@ void UpdateInput(CTxIn& input, const SignatureData& data);
  * Solvability is unrelated to whether we consider this output to be ours. */
 bool IsSolvable(const SigningProvider& provider, const CScript& script);
 
+class SignatureExtractorChecker final : public BaseSignatureChecker
```
I don't think this class is used (or intended to be used) outside of `sign.cpp`, so perhaps move it there, inside the anonymous namespace?

achow101 replied:
Moved

sipa reviewed (commented):
Tiny nits.

achow101 committed:
Make SignatureData able to store signatures and scripts

In addition to having the scriptSig and scriptWitness, have SignatureData
also be able to store just the signatures (pubkeys mapped to sigs) and
scripts (script ids mapped to scripts).

Also have DataFromTransaction be able to extract signatures and scripts
from the scriptSig and scriptWitness of an input to put them in SignatureData.

Adds a new SignatureChecker which takes a SignatureData and puts pubkeys
and signatures into it when it successfully verifies a signature.

Adds a new field in SignatureData which stores whether the SignatureData
was complete. This allows us to also update the scriptSig and
scriptWitness to the final one when updating a SignatureData with another
one.

laanwj commented:
utACK b815600295bd5e7b274c8ee32da65f1419a79ab2
